src/coreclr/jit/importer.cpp

```diff
@@ -3935,29 +3941,19 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
             case NI_System_String_get_Length:
             {
                 GenTree* op1 = impPopStack().val;
-                if (opts.OptimizationEnabled())
+                if (op1->OperIs(GT_CNS_STR))
                 {
-                    if (op1->OperIs(GT_CNS_STR))
+                    // Optimize `ldstr + String::get_Length()` to CNS_INT
+                    // e.g. "Hello".Length => 5
+                    GenTreeIntCon* iconNode = gtNewStringLiteralLength(op1->AsStrCon());
+                    if (iconNode != nullptr)
```

Comment:
Why are we doing this optimization in debug as well now?

Reply:
we don't call this code when OptimizationsEnabled() is false at all, we have a list of sort of "must expand" intrinsics which will, but these won't. Addresses Jan's https://github.com/dotnet/runtime/pull/63734#discussion_r784038670